File: pkg/cfsync/push_live_test.go
// SPDX-FileCopyrightText: (c) 2026 Rafal Zajac
// SPDX-License-Identifier: MIT

//go:build confluence

// These tests hit the live Atlassian Site using the environment loaded by
// liveEnv (see live_test.go). They run only when CFSYNC_TEST_HOST,
// CFSYNC_TEST_ACCOUNT, CFSYNC_TEST_TOKEN and CFSYNC_TEST_SPACE are set (directly
// or via a .env file at the repository root); otherwise they skip.
//
// Run with: go test -tags confluence ./pkg/cfsync/
package cfsync

import (
	"context"
	"path/filepath"
	"testing"

	"github.com/ctx42/testing/pkg/assert"
	"github.com/ctx42/testing/pkg/must"
)

func Test_live_push_noop(t *testing.T) {
	// A pull refreshes a page's Markdown to its baseline, so an immediate push
	// must report no changes and send no update — a full-pipeline smoke test
	// (frontmatter parse, cache load, Put, GetPut) that mutates nothing beyond
	// the throwaway page it creates.

	// --- Given ---
	ctx, client, cfg, lp := liveEnv(t)
	spaceID := must.Value(spaceIDByKey(ctx, client, cfg, lp.space))
	id := must.Value(seedPage(ctx, client, cfg, spaceID, lp.folder,
		uniqueTitle("push-noop"),
		`{"type":"doc","content":[{"type":"paragraph","content":[`+
			`{"type":"text","text":"hello from cfsync"}]}]}`))
	t.Cleanup(func() { _ = purgePage(context.Background(), client, cfg, id) })

	dest := filepath.Join(cfg.WorkDir, "page.md")
	cfg.Pages = map[string]string{
		dest: "/wiki/spaces/" + lp.space + "/pages/" + id + "/it",
	}
	_, _, err := pullPages(ctx, client, cfg)
	must.Nil(err)

	// --- When ---
	have, err := pushPages(ctx, client, cfg, "")

	// --- Then ---
	assert.NoError(t, err)
	assert.Contain(t, "no changes", have)
}

File: packages/cli/test/live/push.live.test.ts
// SPDX-FileCopyrightText: (c) 2026 Rafal Zajac
// SPDX-License-Identifier: MIT

// Ported from push_live_test.go. A pull refreshes a page's Markdown to its
// baseline, so an immediate push must report no changes and send no update.

import { mkdtemp, rm, writeFile } from "node:fs/promises";
import { tmpdir } from "node:os";
import { join } from "node:path";
import {
    afterEach,
    beforeEach,
    describe,
    expect,
    it,
    onTestFinished,
} from "vitest";
import {
    liveConfigured,
    makeRun,
    requireEnv,
    seedClient,
} from "./support/live-env.ts";
import { uniqueTitle } from "./support/probe.ts";

describe.skipIf(!liveConfigured())("live push no-op", () => {
    const env = requireEnv();
    const client = seedClient(env);
    let dir: string;

    beforeEach(async () => {
        dir = await mkdtemp(join(tmpdir(), "cfsync-live-"));
    });
    afterEach(async () => {
        await rm(dir, { recursive: true, force: true });
    });

    it("pull then immediate push reports unchanged", async () => {
        const run = makeRun(env, dir);
        const spaceId = (await client.resolveSpace(env.space)).id;
        const { id } = await client.createPage({
            spaceId,
            title: uniqueTitle("push-noop"),
            parentId: env.folder,
            docJSON:
                '{"type":"doc","content":[{"type":"paragraph","content":[{"type":"text","text":"hello from cfsync"}]}]}',
        });
        onTestFinished(async () => {
            await client.deletePage(id).catch(() => {});
        });

        const cfgPath = join(dir, ".cfsync.yaml");
        await writeFile(
            cfgPath,
            `pages:\n  page.md: /wiki/spaces/${env.space}/pages/${id}/it\n`,
        );

        const pulled = await run(["pull", "--config", cfgPath]);
        expect(pulled.code, pulled.err).toBe(0);

        const pushed = await run(["push", "--config", cfgPath]);
        expect(pushed.code, pushed.err).toBe(0);
        expect(pushed.out).toContain("unchanged");
    });
});
